### ISSUE TITLE
discussion: black and isort for code formatting

### ISSUE THREAD

**Issue by williballenthin** [vivisect/vivisect#294]
Despite having written Python for some years, I've really appreciated using [black](https://github.com/psf/black) and [isort](https://pypi.org/project/isort/) recently. They generally align to my personal style (which is somewhat close to viv, with the exception of snake casing), and they take away all effort required to keep things consistent. Would the viv project consider using these tools to keep all code in a consistent format?

**Comment by rakuy0** [comment]
I'm all for having code in a consistent format so that it all looks, feels, smells, tastes, and gives all glory to the hypnotoad the same. And I like how isort looks. it falls generally in line with how I already like to organize imports. So I'm down for using isort.

Though for most other things, I prefer flake8 or yapf enforcing pep8. There's no strong technical reason for them, I'm just boring and generally like single quotes and how they lay out long lists of function parameters as compared to how black lays them out.

@atlas0fd00m your opinions?

**Comment by atlas0fd00m** [comment]
i like the idea of consistent cleanup.  i'm concerned about tools that are too prescriptive of what that looks like.  for instance, i'm a *huge* fan of camel-case for most things.  my personal style would take a serious hit if i were to have to use snake_case for everything.  i'm also a fan of import-ordering from small to large (Christmas-tree-like), but i also prefer having the Vivisect imports separate from external imports.
can `black` be made to support some sort of normalcy around camel-case for objects/classes/methods?

**Comment by williballenthin** [comment]
black will format the code, but not rename symbols. however, it has a very specific style and there are essentially no options to configure it. in a way, that's part of the charm - to leave the style up to someone else. though, after years of python coding, we all have our ideal style.

isort has a mode to order by line length, and break the imports into groups: stdlib, external, current project.

we've been using this setup for capa and its pretty nice. takes any sort of debate away from how the code should be laid out. example here: https://github.com/fireeye/capa/blob/master/capa/main.py

**Comment by williballenthin** [comment]
bump. with the wide ranging py3 fixes, this is a great time to reformat all the code into a consistent format.

would you accept a PR if i provided a configuration that could be run in GH actions? and/or with all the source reformatted?

**Comment by williballenthin** [comment]
alternatively, please close out this issue if it will not be addressed.

**Comment by rakuy0** [comment]
I agree that it's a good time to reformat things. And I would definitely accept a PR for something to be run in github actions to yell about formatting.

I'm undecided on bluntly have it refactor everything all at once, but I definitely want  to have it scream about making formatting/imports all look the same.

**Comment by atlas0fd00m** [comment]
problem with reformatting everything is that the API's will change.

**Comment by williballenthin** [comment]
@atlas0fd00m can you clarify what you mean? neither `black` nor `isort` will change the semantics or symbols of a program, but they will reformat the layout of text on/across lines.

**Comment by williballenthin** [comment]
> I'm undecided on bluntly have it refactor everything all at once

understood, though with the recent merge of a 45k line PR across almost every file, it seemed now might be a ripe time to do wide ranging cleanup.

**Comment by atlas0fd00m** [comment]
sorry, brainfart.
i was referring to a previous conversation about camel-case versus snake-case, and i thought i was tied into this discussion.  my bad.

**Comment by rakuy0** [comment]
@williballenthin -- Ehh, it's more that I've never been particularly in love with how black formats code (having used it on previous projects). But that's a personal preference that I need to set aside since having an autoformatter is more important. It's not awful, I'm just picky.

I'll go merge #356  unless you have more you want to put in there?

**Comment by williballenthin** [comment]
the action is simply going to fail if the format doesn't match what `black`/`isort` produces. therefore, you'll probably want to do a big reformat. i guess you could reconfigure it to warn, not fail, though i'm don't know how to limit its inspection to just a specific diff, so that probably wont be very helpful either.

**Comment by rakuy0** [comment]
I'm okay with it failing, but yea. I'll throw up a big reformat later once all the current active branches (the latest 4) are merged in.

**Comment by williballenthin** [comment]
sounds great, thank you!
